storage/imx6ull-flashnor/flashnor-qspi.c

```diff
@@ -0,0 +1,302 @@
+	unsigned char jedec[3]; /* Chip JEDEC ID */
+	unsigned int sectorsz;  /* Chip sector size in bytes */
+	unsigned int flashsz;   /* Chip storage size in bytes */
+} chip_t;
+
+/* Supported NOR flash chips */
+static const chip_t chips[] = {
+	{ "Micron MT25QL256ABA(256Mb, 3V)", { 0x20, 0xba, 0x19 }, 4 * 1024, 32 * 1024 * 1024 },
+	/* Not tested.
+
+	{ "Micron MT25QL256ABA(1Gb, 3V)", { 0x20, 0xba, 0x21 }, 64 * 1024, 1024 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(512Mb, 3V)", { 0x20, 0xba, 0x20 }, 64 * 1024, 512 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(128Mb, 3V)", { 0x20, 0xba, 0x18 }, 64 * 1024, 128 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(16Mb, 3V)", { 0x20, 0xba, 0x17 }, 64 * 1024, 16 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(1Gb, 1.8V)", { 0x20, 0xbb, 0x21 }, 64 * 1024, 1024 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(512Mb, 1.8V)", { 0x20, 0xbb, 0x20 }, 64 * 1024, 512 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(256Mb, 1.8V)", { 0x20, 0xbb, 0x19 }, 64 * 1024, 256 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(128Mb, 1.8V)", { 0x20, 0xbb, 0x18 }, 64 * 1024, 128 * 1024 * 1024 },
+	{ "Micron MT25QL256ABA(16Mb, 1.8V)", { 0x20, 0xbb, 0x17 }, 64 * 1024, 16 * 1024 * 1024 },
+	*/
+};
+
+struct {
+	handle_t lock;
+	int init;
+} flashnor_common;
+
+
+static int _flashnor_qspiWriteEnable(qspi_dev_t dev)
+{
+	return _qspi_readBusy(dev, lut_seq_write_enable, 0, NULL, 0);
+}
+
+
+static int _flashnor_qspiWaitBusy(qspi_dev_t dev)
+{
+	uint8_t status;
+	int err, max_iter;
+	unsigned int sleep = 100;
+
+	err = _qspi_readBusy(dev, lut_seq_read_status, 0, &status, 1);
+	if (err < 0) {
+		return err;
+	}
+
+	for (max_iter = 1000; max_iter > 0; max_iter--) {
+		if ((status & 1) == 0) {
+			return EOK;
+		}
+		usleep(sleep);
+		err = _qspi_readBusy(dev, lut_seq_read_status, 0, &status, 1);
+		if (err < 0) {
+			return err;
+		}
+		if (sleep < 100000) {
+			sleep <<= 1;
+		}
+	}
+	return -ETIME;
+}
+
+
+ssize_t flashnor_qspiRead(qspi_dev_t dev, unsigned int addr, void *buff, size_t bufflen)
+{
+	size_t size, len;
+	int err;
+
+	err = mutexLock(flashnor_common.lock);
+	if (err < 0) {
+		return err;
+	}
+
+	for (len = 0; len < bufflen; len += size) {
+		if ((size = bufflen - len) > MAX_READ_LEN)
+			size = MAX_READ_LEN;
+
+		/* TODO XIP */
+		err = _qspi_readBusy(dev, lut_seq_read, addr + len, ((char *)buff) + len, size);
+		if (err < 0) {
+			mutexUnlock(flashnor_common.lock);
+			return err;
+		}
+	}
+
+	err = mutexUnlock(flashnor_common.lock);
+	if (err < 0) {
+		return err;
+	}
+
+	return bufflen;
+}
+
+
+ssize_t flashnor_qspiWrite(qspi_dev_t dev, unsigned int addr, const void *buff, size_t bufflen)
+{
+	size_t size, len;
+	int err;
+
+	err = mutexLock(flashnor_common.lock);
+	if (err < 0) {
+		return err;
+	}
+
+	for (len = 0; len < bufflen; len += size) {
+		err = _flashnor_qspiWriteEnable(dev);
+		if (err < 0) {
+			mutexUnlock(flashnor_common.lock);
+			return err;
+		}
+		if ((size = bufflen - len) > MAX_WRITE_LEN)
+			size = MAX_WRITE_LEN;
```

Comment:
MISRA rule 13.4 The result of an assignment operator should not be used

```suggestion
		size = bufflen - len;
		if (size > MAX_WRITE_LEN) {
			size = MAX_WRITE_LEN;
		}
```